### PR TITLE
Refactor pinned memory vector and ORC+Parquet writers

### PR DESCRIPTION
**Pull request by ttnghia**
Currently, `cudf::detail::pinned_allocator` is used in various places to implement a pinned host vector. This standardizes such usage, removing `cudf::detail::pinned_allocator` from the usage sites and replacing its usage by a standard `cudf::detail::pinned_host_vector` instead.

Some small changes are also made for ORC/Parquet writer classes, replacing `bool _single_write_mode` by `SingleWriteMode _single_write_mode`.

This is needed for https://github.com/rapidsai/cudf/pull/13076.

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/pinned_host_vector.hpp
Comment (PointKernel) [inline cpp/include/cudf/detail/utilities/pinned_host_vector.hpp:1]: IIUC, pinned memory is always on the host side thus not sure this renaming is really needed.

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/src/io/orc/writer_impl.cu [hunk 1]
```diff
@@ -379,11 +375,11 @@ __global__ void copy_string_data(char* string_pool,
 }  // namespace
 
 void persisted_statistics::persist(int num_table_rows,
-                                   bool single_write_mode,
+                                   SingleWriteMode single_write_mode,
```
Comment (PointKernel) [inline cpp/src/io/orc/writer_impl.cu:378]: According to https://github.com/rapidsai/cudf/blob/branch-23.06/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#code-and-documentation-style-and-formatting, We use snake_case to name enums.

File: cpp/src/io/orc/writer_impl.cu [hunk 2]
```diff
@@ -379,11 +375,11 @@ __global__ void copy_string_data(char* string_pool,
...
                                    intermediate_statistics& intermediate_stats,
                                    rmm::cuda_stream_view stream)
 {
-  if (not single_write_mode) {
+  if (single_write_mode == SingleWriteMode::YES) {
```
Comment (PointKernel) [inline cpp/src/io/orc/writer_impl.cu:382]:
```suggestion
  if (single_write_mode == SingleWriteMode::NO) {
```
Seems to be `NO` here.

File: cpp/src/io/orc/writer_impl.cu [hunk 3]
```diff
@@ -79,11 +80,6 @@ struct row_group_index_info {
 };
 
 namespace {
-/**
- * @brief Helper for pinned host memory
- */
-template <typename T>
-using pinned_buffer = std::unique_ptr<T, decltype(&cudaFreeHost)>;
```
Comment (vuule) [inline cpp/src/io/orc/writer_impl.cu:86]: This looks good, with one minor concern. For places that already used `host_vector`, there's no change. But, for this use case, we are introducing initialization into the memory that does not need to be initialized (and previously wasn't). If you don't mind, please run ORC or Parquet writer benchmarks.

Comment (ttnghia) [reply cpp/src/io/orc/writer_impl.cu:86]: Sorry I didn't see where is the initialization that you mentioned? The new `pinned_host_vector` uses allocator that also doesn't initialize the internal buffer:
```
__host__ inline pointer allocate(size_type cnt, const_pointer /*hint*/ = 0)
  {
    if (cnt > this->max_size()) { throw std::bad_alloc(); }  // end if

    pointer result(0);
    CUDF_CUDA_TRY(cudaMallocHost(reinterpret_cast<void**>(&result), cnt * sizeof(value_type)));
    return result;
  }
```

**Comment by ttnghia** [comment]
@vuule Here is some bechmarks for ORC writer:
```
['./before.json', 'after.json']
# orc_write_encode

## [0] Quadro RTX 6000

|    data_type    |  cardinality  |  run_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|-----------------|---------------|--------------|------------|-------------|------------|-------------|--------------|---------|----------|
| INTEGRAL_SIGNED |       0       |      1       | 127.838 ms |       3.63% | 126.039 ms |       2.71% | -1799.682 us |  -1.41% |   PASS   |
| INTEGRAL_SIGNED |     1000      |      1       | 254.980 ms |       1.54% | 257.532 ms |       1.94% |     2.552 ms |   1.00% |   PASS   |
| INTEGRAL_SIGNED |       0       |      32      |  52.873 ms |       3.76% |  54.257 ms |       4.01% |     1.385 ms |   2.62% |   PASS   |
| INTEGRAL_SIGNED |     1000      |      32      |  55.094 ms |       1.51% |  55.554 ms |       2.56% |   459.615 us |   0.83% |   PASS   |
|     STRING      |       0       |      1       | 545.909 ms |       0.02% | 547.088 ms |       0.37% |     1.178 ms |   0.22% |   FAIL   |
|     STRING      |     1000      |      1       | 813.042 ms |       0.11% | 813.765 ms |       0.15% |   722.949 us |   0.09% |   PASS   |
|     STRING      |       0       |      32      | 545.776 ms |       0.07% | 546.026 ms |       0.10% |   250.000 us |   0.05% |   PASS   |
|     STRING      |     1000      |      32      | 650.413 ms |       0.07% | 650.709 ms |       0.12% |   295.630 us |   0.05% |   PASS   |
|      LIST       |       0       |      1       | 133.945 ms |       0.31% | 134.238 ms |       0.35% |   293.109 us |   0.22% |   PASS   |
|      LIST       |     1000      |      1       | 336.120 ms |       0.10% | 337.486 ms |       0.82% |     1.367 ms |   0.41% |   FAIL   |
|      LIST       |       0       |      32      |  88.915 ms |       0.35% |  88.975 ms |       0.40% |    60.099 us |   0.07% |   PASS   |
|      LIST       |     1000      |      32      | 104.201 ms |       0.34% | 104.000 ms |       0.30% |  -201.286 us |  -0.19% |   PASS   |
|     STRUCT      |       0       |      1       | 547.717 ms |       0.16% | 548.105 ms |       0.15% |   387.793 us |   0.07% |   PASS   |
|     STRUCT      |     1000      |      1       | 902.261 ms |       0.43% | 901.978 ms |       0.08% |  -282.324 us |  -0.03% |   PASS   |
|     STRUCT      |       0       |      32      | 527.460 ms |       0.08% | 528.633 ms |       0.25% |     1.173 ms |   0.22% |   FAIL   |
|     STRUCT      |     1000      |      32      | 607.662 ms |       0.27% | 608.766 ms |       0.52% |     1.103 ms |   0.18% |   PASS   |

# orc_write

## [0] Quadro RTX 6000

|  num_columns  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|---------------|------------|-------------|------------|-------------|------------|---------|----------|
|       8       | 165.981 ms |       0.30% | 166.784 ms |       0.49% | 803.009 us |   0.48% |   FAIL   |
|      64       |    1.558 s |       1.41% |    1.591 s |       3.23% |  32.855 ms |   2.11% |   FAIL   |

# orc_chunked_write

## [0] Quadro RTX 6000

|  num_columns  |  num_chunks  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|---------------|--------------|------------|-------------|------------|-------------|---------------|---------|----------|
|       8       |      8       | 393.974 ms |       0.51% | 425.155 ms |       2.94% |     31.181 ms |   7.91% |   FAIL   |
|      64       |      8       |    3.969 s |        inf% |    4.332 s |        inf% |    362.998 ms |   9.15% |   PASS   |
|       8       |      64      |    1.831 s |       0.12% |    1.976 s |       0.95% |    144.625 ms |   7.90% |   FAIL   |
|      64       |      64      |    7.401 s |        inf% |    7.366 s |        inf% | -34970.378 us |  -0.47% |   PASS   |

```

File: cpp/src/io/orc/writer_impl.cu (hunk 3 above)
Comment (ttnghia) [reply cpp/src/io/orc/writer_impl.cu:86]: Please see benchmarks here: https://github.com/rapidsai/cudf/pull/13206#issuecomment-1520782754

Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:86]: I thought that vector has its own initialization outside of the allocator. Either way, it does not seem to impact the overall performance.

**Review by vuule — APPROVED** [review]
Thank you for taking the time to make these improvements.

File: cpp/include/cudf/detail/utilities/pinned_host_vector.hpp
Comment (ttnghia) [reply cpp/include/cudf/detail/utilities/pinned_host_vector.hpp:1]: IMO the name `host_vector` is better expressing its purpose, similar to `thurst::host_vector` vs `thrust::vector`.

**Review by PointKernel — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/pinned_host_vector.hpp
Comment (vuule) [reply cpp/include/cudf/detail/utilities/pinned_host_vector.hpp:1]: I read the name as host_vector in pinned memory, so the name looks good.

**Pull request merged**